crates/lang/ir/src/ir/attrs.rs

```diff
@@ -723,13 +759,64 @@ impl From<InkAttribute> for Attribute {
     }
 }
 
+/// This function replaces occurrences of a `TokenTree::Ident` of the verbatim `_`
+/// with the String literal `"_"`.
+///
+/// This is done because `syn::Attribute::parse_meta` does not support parsing
+/// verbatim like `_`. For this we would need to switch to `syn::Attribute::parse_args`,
+/// which requires a more in-depth rewrite of our IR parsing.
+fn transform_wildcard_selector_to_string(grp: Group2) -> TokenTree2 {
+    let mut found_selector = false;
+    let mut found_equal = false;
+
+    let new_grp: TokenStream2 = grp
+        .stream()
+        .into_iter()
+        .map(|tt| {
+            match tt {
+                TokenTree2::Group(grp) => transform_wildcard_selector_to_string(grp),
+                TokenTree2::Ident(ident)
+                    if found_selector && found_equal && ident == "_" =>
+                {
+                    let mut lit = proc_macro2::Literal::string("_");
+                    lit.set_span(ident.span());
+                    TokenTree2::Literal(lit)
+                }
+                TokenTree2::Ident(ident) if ident == "selector" => {
+                    found_selector = true;
+                    TokenTree2::Ident(ident)
+                }
+                TokenTree2::Punct(punct) if punct.as_char() == '=' => {
+                    found_equal = true;
+                    TokenTree2::Punct(punct)
+                }
+                _ => tt,
+            }
+        })
+        .collect();
+    TokenTree2::Group(Group2::new(grp.delimiter(), new_grp))
+}
+
 impl TryFrom<syn::Attribute> for InkAttribute {
     type Error = syn::Error;
 
-    fn try_from(attr: syn::Attribute) -> Result<Self, Self::Error> {
+    fn try_from(mut attr: syn::Attribute) -> Result<Self, Self::Error> {
         if !attr.path.is_ident("ink") {
             return Err(format_err_spanned!(attr, "unexpected non-ink! attribute"))
         }
+
+        let ts: TokenStream2 = attr
+            .tokens
+            .into_iter()
+            .map(|tt| {
+                match tt {
+                    TokenTree2::Group(grp) => transform_wildcard_selector_to_string(grp),
+                    _ => tt,
+                }
+            })
+            .collect();
```

Comment:
This is very fragile because it changes syntax without looking at the semantics.
A point where this can fail is any input that includes a `_` token, such as:
```rust
#[compose(Bar, _)]
#[precondition x == Some(_)]
```

Since with this implementation we accept `_` as well as `"_"` we should just remove this transformation and instead use a custom parsing local to this code point that parses the given `attr.tokens` instead of blindfoldedly applying these generic transformations to its input. And if the parsing finds `selector = _` then it can replace it with `selector = "_"`. This way we have a stable implementation without requiring too much efforts into re-implementing the custom parsing for attributes everywhere.

Reply:
Yeah it's fragile, the case you mentioned is not possible though, since the impl of `transform_wildcard_selector_to_string` already has two preconditions that match on `selector = ` before replacing.